internal/migrations/migrator.go

```diff
@@ -0,0 +1,255 @@
+	// TODO: get changeset
+	var cs Changeset
+	err := m.storeChangeset(&cs)
+	if err != nil {
+		return err
+	}
+
+	m.lastChangeset = block.Height
+	return nil
+}
+
+var ErrNoActiveMigration = fmt.Errorf("no active migration")
+
+// MigrationMetadata holds metadata about a migration, informing
+// consumers of what information the current node has available
+// for the migration.
+type MigrationMetadata struct {
+	// GenesisSnapshot holds information about the genesis snapshot.
+	GenesisSnapshot *statesync.Snapshot
+	// LastChangeset is the height of the last changeset that was stored.
+	// Nodes are expected to have all changesets from LastChangeset to
+	// Snapshot.Height. If LastChangeset is -1, then no changesets have
+	// been stored yet.
+	LastChangeset int64
+}
+
+// GetMigrationMetadata gets the metadata for the genesis snapshot,
+// as well as the available changesets.
+func (m *Migrator) GetMigrationMetadata() (*MigrationMetadata, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	if m.activeMigration == nil {
+		return nil, ErrNoActiveMigration
+	}
+
+	snapshots := m.snapshotter.ListSnapshots()
+	if len(snapshots) == 0 {
+		return nil, fmt.Errorf("migration is active, but no snapshots found. The node might still be creating the snapshot")
+	}
+	if len(snapshots) > 1 {
+		return nil, fmt.Errorf("migration is active, but more than one snapshot found. This should not happen, and is likely a bug")
+	}
+
+	return &MigrationMetadata{
+		GenesisSnapshot: snapshots[0],
+		LastChangeset:   m.lastChangeset,
+	}, nil
+}
+
+// GetGenesisChunk gets the genesis chunk for the migration.
+func (m *Migrator) GetGenesisChunk(height int64, format uint32, chunkIdx uint32) ([]byte, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	if m.activeMigration == nil {
+		return nil, ErrNoActiveMigration
+	}
+	return m.snapshotter.LoadSnapshotChunk(uint64(height), format, chunkIdx)
+}
+
+// GetChangeset gets the changeset for a block in the migration.
+func (m *Migrator) GetChangeset(height int64) (*Changeset, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	if m.activeMigration == nil {
+		return nil, ErrNoActiveMigration
+	}
+	if height < m.activeMigration.StartHeight {
+		return nil, fmt.Errorf("requested changeset height is before the start of the migration")
+	}
+	if height > m.activeMigration.EndHeight {
+		return nil, fmt.Errorf("requested changeset height is after the end of the migration")
+	}
+	if height > m.lastChangeset {
+		return nil, fmt.Errorf("requested changeset height has not been recorded by the node yet")
+	}
+
+	return m.loadChangeset(height)
+}
+
+// storeChangeset persists a changeset to the migration directory.
+func (m *Migrator) storeChangeset(c *Changeset) error {
+	bts, err := c.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Create(filepath.Join(m.dir, formatChangesetFilename(c.Height)))
+	if err != nil {
+		return err
+	}
+
+	_, err = file.Write(bts)
```

Comment:
Yeah this entire section is going to change, since we are going to incrementally writing changesets to a file with an `io.Writer`. I made this section, then got into Postgres, and then have yet to circle back.